src/types/mappingproxy.rs

```diff
@@ -0,0 +1,597 @@
+// Copyright (c) 2017-present PyO3 Project and Contributors
+
+use super::PyMapping;
+use crate::err::PyResult;
+use crate::types::dict::{IntoPyDict, PyDictItem};
+use crate::types::{PyAny, PyIterator, PyList, PySequence};
+use crate::{ffi, AsPyPointer, PyErr, PyTryFrom, Python, ToPyObject};
+
+/// Represents a Python `mappingproxy`.
+#[repr(transparent)]
+pub struct PyMappingProxy(PyAny);
+
+pyobject_native_type_core!(
+    PyMappingProxy,
+    pyobject_native_static_type_object!(ffi::PyDictProxy_Type),
+    #checkfunction=ffi::PyDictProxy_Check
+);
+
+impl PyMappingProxy {
+    /// Creates a mappingproxy from an object.
+    pub fn new<'py>(py: Python<'py>, elements: &'py PyMapping) -> PyResult<&'py PyMappingProxy> {
+        unsafe {
+            let proxy = ffi::PyDictProxy_New(elements.as_ptr());
+            py.from_owned_ptr_or_err::<PyMappingProxy>(proxy)
+        }
+    }
+
+    /// Returns a new mappingproxy that contains the same key-value pairs as self.
+    ///
+    /// This is equivalent to the Python expression `self.copy()`.
+    pub fn copy(&self) -> PyResult<&PyMapping> {
+        self.call_method0("copy")
+            .and_then(|object| object.downcast::<PyMapping>().map_err(PyErr::from))
+    }
+
+    /// Checks if the mappingproxy is empty, i.e. `len(self) == 0`.
+    pub fn is_empty(&self) -> bool {
+        self.len().unwrap_or_default() == 0
+    }
+
+    /// Gets an item from the mappingproxy.
+    ///    
+    /// Returns `None` if the item is not present, or if an error occurs.
+    ///
+    /// To get a `KeyError` for non-existing keys, use `PyAny::get_item`.
+    pub fn get_item<K>(&self, key: K) -> Option<&PyAny>
+    where
+        K: ToPyObject,
+    {
+        PyAny::get_item(self, key).ok()
+    }
+
+    /// Returns a list of mappingproxy keys.
+    ///
+    /// This is equivalent to the Python expression `list(mappingproxy.keys())`.
+    pub fn keys(&self) -> &PyList {
+        unsafe {
+            PySequence::try_from_unchecked(self.call_method0("keys").unwrap())
+                .to_list()
+                .unwrap()
+        }
+    }
```

Comment:
The unwraps are not ok here, let's have `PyResult<&PyList>` as the return type please!